File: Data Structures & Algorithms/top-k-elements-in-list/submission-0.py
from collections import defaultdict

class Solution:

    # INPUT: int arr (nums), int k
    # OUTPUT: k most frequent elements in arr (in any order -> one solution)

    # Edge Cases:
    # 1. arr is empty, one element -> return the element
    # 2. k is 0, return []

    # Plan
    #  (loop through every elem at least once)
    # 1. Create a default dict (key -> number, value -> num of occurrences)
    # 2. Loop through the list, and update the dict accordingly
    # 3. Sort the dict (based on the values)
    # 3b. Returning the top k elements

    # ISSUE: {a -> 3, b -> 4, c -> 3 ...}

    # 1. Do an initial pass and get num of unique elements (using a hashset)
    # 2. Create an arr of len Unique, 

    # 1. Create Counter dict
    # 2. Sort by value, and get the key as an array
    # 3. Return the first k elements

    # Time: O(n log n). Space: O(n)
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = defaultdict(int)

        for elem in nums:
            count[elem] += 1
        
        count = [number for number, occurrence in sorted(count.items(), key = lambda item: item[1], reverse = True)]

        return count[:k]

File: Data Structures & Algorithms/binary-search/submission-0.kt
class Solution {

    // Input: arr of nums (sorted), target (int)
    // Output: Index of the number in the array or -1 if absent

    // Edge Case: Empty array -> -1

    // PLAN:
    // 1a. Create left and right indices
    // 1. Check the middle, compare it with the target ([while] looping with the index, number)
    // 2. If middle == target, return the index;
    //       middle > target, move right = middle - 1
    //       middle < target, move left = left + 1
    // 3. If left > right, return -1  

    // Time: O(log n). Space: O(1)
    fun search(nums: IntArray, target: Int): Int {
        var left = 0
        var right = nums.size - 1

        if (nums.size == 0) return -1

        while (left <= right) {
            val middleIndex = (left + right) / 2
            val middle = nums[middleIndex]
            
            if (middle == target)
                return middleIndex
            else if (middle > target) {
                right = middleIndex - 1
            }
            else {
                left = middleIndex + 1
            }
        }

        return -1
    }
}
